Userland/Libraries/LibSoftGPU/Clipper.cpp

```diff
@@ -64,42 +69,94 @@ static constexpr GPU::Vertex clip_intersection_point(GPU::Vertex const& p1, GPU:
     return out;
 }
 
-template<Clipper::ClipPlane plane>
-FLATTEN static constexpr void clip_plane(Vector<GPU::Vertex>& input_list, Vector<GPU::Vertex>& output_list)
+template<Clipper::ClipFrustrum plane>
+FLATTEN static constexpr void clip_frustrum(Vector<GPU::Vertex>& input_list, Vector<GPU::Vertex>& output_list)
+{
+    constexpr FloatVector4 clip_plane_normals[] = {
+        { 1, 0, 0, 1 },  // Left Plane
+        { -1, 0, 0, 1 }, // Right Plane
+        { 0, -1, 0, 1 }, // Top Plane
+        { 0, 1, 0, 1 },  // Bottom plane
+        { 0, 0, 1, 1 },  // Near Plane
+        { 0, 0, -1, 1 }  // Far Plane
+    };
+    constexpr auto clip_plane_normal = clip_plane_normals[to_underlying(plane)];
+
+    output_list.clear_with_capacity();
+
+    auto input_list_size = input_list.size();
+    if (input_list_size == 0)
+        return;
+
+    auto const* prev_vec = &input_list.data()[0];
+    auto is_prev_point_within_frustrum = point_within_frustrum<plane>(prev_vec->clip_coordinates);
+
+    for (size_t i = 1; i <= input_list_size; i++) {
+        auto const& curr_vec = input_list[i % input_list_size];
+        auto const is_curr_point_within_frustrum = point_within_frustrum<plane>(curr_vec.clip_coordinates);
+
+        if (is_curr_point_within_frustrum != is_prev_point_within_frustrum)
+            output_list.append(clip_intersection_point<Clipper::Space::CLIP>(*prev_vec, curr_vec, clip_plane_normal));
+
+        if (is_curr_point_within_frustrum)
+            output_list.append(curr_vec);
+
+        prev_vec = &curr_vec;
+        is_prev_point_within_frustrum = is_curr_point_within_frustrum;
+    }
+}
```

Comment:
This entire method is duplicated in `clip_plane` below, why not template this method and call either `point_within_frustum` or `point_within_plane` based on an extra template parameter?

Reply:
As suggested, my latest commit consolidates all of the duplicate methods, instead relying on templates.